server/users.go

```diff
@@ -1,317 +1,153 @@
-	return srcID, store, nil
-}
-
-// hasRoles checks if the influx source has roles or not
-func (h *Service) hasRoles(ctx context.Context, ts chronograf.TimeSeries) (chronograf.RolesStore, bool) {
-	store, err := ts.Roles(ctx)
-	if err != nil {
-		return nil, false
+	if err := s.UsersStore.Delete(ctx, u); err != nil {
+		Error(w, http.StatusBadRequest, err.Error(), s.Logger)
 	}
-	return store, true
-}
 
-type userRequest struct {
-	Username    string                 `json:"name,omitempty"`        // Username for new account
-	Password    string                 `json:"password,omitempty"`    // Password for new account
-	Permissions chronograf.Permissions `json:"permissions,omitempty"` // Optional permissions
-	Roles       []chronograf.Role      `json:"roles,omitempty"`       // Optional roles
+	w.WriteHeader(http.StatusNoContent)
 }
 
-func (r *userRequest) ValidCreate() error {
-	if r.Username == "" {
-		return fmt.Errorf("Username required")
-	}
-	if r.Password == "" {
-		return fmt.Errorf("Password required")
+func (s *Service) UpdateUser(w http.ResponseWriter, r *http.Request) {
+	var req userRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		invalidJSON(w, s.Logger)
+		return
 	}
-	return validPermissions(&r.Permissions)
-}
 
-type usersResponse struct {
-	Users []userResponse `json:"users"`
-}
-
-func (r *userRequest) ValidUpdate() error {
-	if r.Password == "" && len(r.Permissions) == 0 && len(r.Roles) == 0 {
-		return fmt.Errorf("No fields to update")
+	if err := req.ValidUpdate(); err != nil {
+		invalidData(w, err, s.Logger)
+		return
 	}
-	return validPermissions(&r.Permissions)
-}
 
-type userResponse struct {
-	Name           string                 // Username for new account
-	Permissions    chronograf.Permissions // Account's permissions
-	Roles          []roleResponse         // Roles if source uses them
-	Links          selfLinks              // Links are URI locations related to user
-	hasPermissions bool
-	hasRoles       bool
-}
-
-func (u *userResponse) MarshalJSON() ([]byte, error) {
-	res := map[string]interface{}{
-		"name":  u.Name,
-		"links": u.Links,
-	}
-	if u.hasRoles {
-		res["roles"] = u.Roles
-	}
-	if u.hasPermissions {
-		res["permissions"] = u.Permissions
-	}
-	return json.Marshal(res)
-}
+	ctx := r.Context()
+	id := httprouter.GetParamFromContext(ctx, "id")
 
-// newUserResponse creates an HTTP JSON response for a user w/o roles
-func newUserResponse(srcID int, name string) *userResponse {
-	self := newSelfLinks(srcID, "users", name)
-	return &userResponse{
-		Name:  name,
-		Links: self,
+	u, err := s.UsersStore.Get(ctx, id)
+	if err != nil {
+		Error(w, http.StatusNotFound, err.Error(), s.Logger)
 	}
-}
 
-func (u *userResponse) WithPermissions(perms chronograf.Permissions) *userResponse {
-	u.hasPermissions = true
-	if perms == nil {
-		perms = make(chronograf.Permissions, 0)
-	}
-	u.Permissions = perms
-	return u
-}
+	// TODO: should we diff and only set non-nil fields?
```

Comment:
Yes, you should only diff.  There are several styles for handling PATCH in REST interfaces, but, the one we've been using for chronograf is to only change the fields that have updates.

In this case if a request didn't set the Name, then the Name would be an empty string.

Reply:
cool - thanks